zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/metrics/ExecutionLatencyMetrics.java

```diff
@@ -7,86 +7,100 @@
-      Histogram.build()
-          .namespace("zeebe")
-          .name("job_life_time")
-          .help("The life time of an job")
-          .labelNames("partition")
-          .buckets(0.10f, 0.2, 0.4, 0.8, 1.6, 3.2, 6.4, 12.8, 25.6, 51.2)
-          .buckets(0.025, 0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1.0, 2.5, 5.0, 10.0, 15.0, 30.0, 45.0)
-          .register();
-  private static final Histogram JOB_ACTIVATION_TIME =
-      Histogram.build()
-          .namespace("zeebe")
-          .name("job_activation_time")
-          .help("The time until an job was activated")
-          .labelNames("partition")
-          .buckets(0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1.0, 2.5, 5.0, 10.0, 15.0, 30.0)
-          .register();
-  private static final Gauge CURRENT_CACHED_INSTANCE =
-      Gauge.build()
-          .namespace("zeebe")
-          .name("execution_latency_current_cached_instances")
-          .help(
-              "The current cached instances for counting their execution latency. If only short-lived instances are handled this can be seen or observed as the current active instance count.")
-          .labelNames("partition", "type")
-          .register();
+  private static final Duration[] JOB_LIFE_TIME_BUCKETS =
+      Stream.of(25, 50, 75, 100, 250, 500, 750, 1000, 2500, 5000, 10000, 15000, 30000, 45000)
+          .map(Duration::ofMillis)
+          .toArray(Duration[]::new);
+
+  private static final Duration[] JOB_ACTIVATION_TIME_BUCKETS =
+      Stream.of(10, 25, 50, 75, 100, 250, 500, 750, 1000, 2500, 5000, 10000, 15000, 30000)
+          .map(Duration::ofMillis)
+          .toArray(Duration[]::new);
+
+  private static final Duration[] PROCESS_INSTANCE_EXECUTION_BUCKETS =
+      Stream.of(50, 75, 100, 250, 500, 750, 1000, 2500, 5000, 10000, 15000, 30000, 45000, 60000)
+          .map(Duration::ofMillis)
+          .toArray(Duration[]::new);
+
+  private final MeterRegistry meterRegistry;
+  private final Map<Integer, AtomicInteger> currentCachedInstanceJobsCount =
+      new ConcurrentHashMap<>();
+  private final Map<Integer, AtomicInteger> currentCacheInstanceProcessInstances =
+      new ConcurrentHashMap<>();
+
+  public ExecutionLatencyMetrics() {
+    this(new SimpleMeterRegistry());
+  }
+
+  public ExecutionLatencyMetrics(final MeterRegistry meterRegistry) {
+    this.meterRegistry = meterRegistry;
+  }
 
   public void observeProcessInstanceExecutionTime(
       final int partitionId, final long creationTimeMs, final long completionTimeMs) {
-    PROCESS_INSTANCE_EXECUTION
-        .labels(Integer.toString(partitionId))
-        .observe(latencyInSeconds(creationTimeMs, completionTimeMs));
+    Timer.builder("zeebe.process.instance.execution.time")
+        .description("The execution time of processing a complete process instance")
+        .tag("partition", Integer.toString(partitionId))
+        .sla(PROCESS_INSTANCE_EXECUTION_BUCKETS)
+        .register(meterRegistry)
```

Comment:
There is no way to tag a meter dynamically and ALSO provide a description and set the sla buckets. There is only the for example `meterRegistry.timer(name, <tags>)` which will create a timer in the registry, to get a description and set the buckets it will require setting the meter registry config such that these get attached when a specific timer name gets created.

The reason this is done here differently than the `ElasticSearchMetrics` where its created in the constructor, is that the ELS metrics does not need to dynamically tag the partitionId, there it is provided in the constructor and all metrics recorded against that instance will use the partitionId in the constructor.

However here the functions which observe metrics also take a partitionId which in micrometer means a new object (only if the timer doesn't exist the documentation says `The returned timer will be unique for each registry, but each registry is guaranteed to only create one timer for the same combination of name and tags.`

Therefore my understanding is they are only created the first time a set of metric name and tags is observed.

Reply:
I see. This is a holdover from the past where the exporter did not know the partition ID from the get go, and read it from the record during exporting. This could actually be refactored now, since we already know the partition ID in the configure method of the exporter, so we could create the metrics class at that point and have no dynamic tags.

Considering it's a small change, I would do it in this PR, but I'm fine with doing it as a follow up.

Reply:
> ❌ Why not create these objects in the constructor to avoid having to recreate them here every time?

Also not sure what is meant my this, they are not recreated every time only the first time during creation. Am I misunderstanding something?